src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,368 @@
+  uint64_t _prng_seed;
+  int _node_count;
+
+  uint64_t prng_next() {
+    // Taken directly off of JFRPrng
+    static const constexpr uint64_t PrngMult = 0x5DEECE66DLL;
+    static const constexpr uint64_t PrngAdd = 0xB;
+    static const constexpr uint64_t PrngModPower = 48;
+    static const constexpr uint64_t PrngModMask = (static_cast<uint64_t>(1) << PrngModPower) - 1;
+    _prng_seed = (PrngMult * _prng_seed + PrngAdd) & PrngModMask;
+    return _prng_seed;
+  }
+
+  struct node_pair {
+    TreapNode* left;
+    TreapNode* right;
+  };
+
+  enum SplitMode {
+    LT, // <
+    LEQ // <=
+  };
+
+  // Split tree at head into two trees, SplitMode decides where EQ values go.
+  // We have SplitMode because it makes remove() trivial to implement.
+  static node_pair split(TreapNode* head, const K& key, SplitMode mode = LEQ DEBUG_ONLY(COMMA int recur_count = 0)) {
+    assert(recur_count < 200, "Call-stack depth should never exceed 200");
+
+    if (head == nullptr) {
+      return {nullptr, nullptr};
+    }
+    if ((COMPARATOR::cmp(head->_key, key) <= 0 && mode == LEQ) || (COMPARATOR::cmp(head->_key, key) < 0 && mode == LT)) {
+      node_pair p = split(head->_right, key, mode DEBUG_ONLY(COMMA recur_count + 1));
+      head->_right = p.left;
+      return node_pair{head, p.right};
+    } else {
+      node_pair p = split(head->_left, key, mode DEBUG_ONLY(COMMA recur_count + 1));
+      head->_left = p.right;
+      return node_pair{p.left, head};
+    }
+  }
+
+  // Invariant: left is a treap whose keys are LEQ to the keys in right.
+  static TreapNode* merge(TreapNode* left, TreapNode* right DEBUG_ONLY(COMMA int recur_count = 0)) {
+    assert(recur_count < 200, "Call-stack depth should never exceed 200");
+
+    if (left == nullptr) return right;
+    if (right == nullptr) return left;
+
+    if (left->_priority > right->_priority) {
+      // We need
+      //      LEFT
+      //         |
+      //         RIGHT
+      // for the invariant re: priorities to hold.
+      left->_right = merge(left->_right, right DEBUG_ONLY(COMMA recur_count + 1));
+      return left;
+    } else {
+      // We need
+      //         RIGHT
+      //         |
+      //      LEFT
+      // for the invariant re: priorities to hold.
+      right->_left = merge(left, right->_left DEBUG_ONLY(COMMA recur_count + 1));
+      return right;
+    }
+  }
+
+  static TreapNode* find(TreapNode* node, const K& k DEBUG_ONLY(COMMA int recur_count = 0)) {
+    if (node == nullptr) {
+      return nullptr;
+    }
+
+    int key_cmp_k = COMPARATOR::cmp(node->key(), k);
+
+    if (key_cmp_k == 0) { // key EQ k
+      return node;
+    }
+
+    if (key_cmp_k < 0) { // key LT k
+      return find(node->right(), k DEBUG_ONLY(COMMA recur_count + 1));
+    } else { // key GT k
+      return find(node->left(), k DEBUG_ONLY(COMMA recur_count + 1));
+    }
+  }
+
+#ifdef ASSERT
+  void verify_self() {
+    const double expected_maximum_depth = log(this->_node_count+1) * 5;
```

Comment:
Where did 5 come from, shouldn't this be:

`    const double expected_maximum_depth = log(this->_node_count+1) * (this->_node_count+1)
`
?

Reply:
The depth of a binary tree is on the order of `log(n)`, "the order of" is important here. Essentially, we need some wiggle room. I found that 3 fails, so I bumped it to 5. This did cause me to investigate whether we can pick a tighter bound, and `3.5` fits as long as we perform `ceil` instead of `floor` on the result.

I'll comment where the constant comes from.

Reply:
This made me a bit curious about other trees bounds. An RB-tree has a bound of `2log_2(N + 1)`, so I decided to find the bound for our treap in base-2 log. Turns out we're at `2.5*log_2(N + 1)`, so very close to the RB-tree.

Reply:
>#  assert(maximum_depth_found <= ceil(expected_maximum_depth)) failed: depth unexpectedly large, was: 57, expected: 56

Apparently there's a very small chance of it being worse than that :-). I'll adjust the expectations here a bit more.

Reply:
The absolute worst case scenario is O(n) and the best is O(log(n)) and since we use random numbers to balance the tree it should be closer to the best case, correct?

I'm just wondering whether we really want:

`    assert(maximum_depth_found <= (int)expected_maximum_depth, "depth unexpectedly large");
`

We cannot guarantee that it will not be triggered at some point in the future, and if it does, how useful it will be for someone to investigate? I would consider printing out a warning instead of using `assert` here.

Reply:
>The absolute worst case scenario is O(n) and the best is O(log(n)) and since we use random numbers to balance the tree it should be closer to the best case, correct?

Right. In fact, it shouldn't be even close to O(n).

 >We cannot guarantee that it will not be triggered at some point in the future, and if it does, how useful it will be for someone to investigate? I would consider printing out a warning instead of using assert here.

To be clear: This is in a verification function only called in two specific tests meant to challenge the treap (in some sense). If we do have a large deviation then that is probably worth investigating, it's an indication that something is going on. If we have intermittent failures because we chose a very tight bound, then we might have to increase that bound slightly, or we investigate how good our PRNG is.